src/hooks/SocialAuth.ts

```diff
@@ -0,0 +1,104 @@
+import { randomBytes } from 'crypto'
+import useLocalStorage from 'hooks/LocalStorage'
+import usePopupWindow from 'hooks/PopupWindow'
+import { useEffect, useState } from 'react'
+
+type OAuthData = {
+  code: string
+}
+
+type OauthOptions = {
+  client_id: string
+  scope: string
+  response_type?: 'code' | 'token'
+  code_challenge?: 'challenge'
+  code_challenge_method?: 'plain'
+}
+
+const useOauthPopupWindow = (url: string, oauthOptions: OauthOptions) => {
+  const [csrfToken, setCsrfToken] = useLocalStorage(
+    `oauth_csrf_token_${oauthOptions.client_id}`,
+    randomBytes(16).toString('hex'),
+    true,
+  )
+
+  const redirectUri = process.env.NEXT_PUBLIC_TWITTER_REDIRECT_URI as string
+
+  oauthOptions.response_type = oauthOptions.response_type ?? 'code'
+
+  const state = `${oauthOptions.client_id};${csrfToken}`
+
+  // prettier-ignore
+  const { onOpen, windowInstance } = usePopupWindow(
+    `${url}?${Object.entries(oauthOptions).map(([key, value]) => `${key}=${value}`).join("&")}&redirect_uri=${encodeURIComponent(redirectUri)}&state=${encodeURIComponent(state)}`
+  )
```

Comment:
**suggestion** i'd suggest we can add `qs` as a dependency and use it here. We need it in a few other spots in the app too.

https://www.npmjs.com/package/qs